src/app/modules/payments/stripe.utils.ts

```diff
@@ -0,0 +1,256 @@
+import config from '../../config/config'
+import { createLoggerWithLabel } from '../../config/logger'
+
+import {
+  ComputePaymentStateError,
+  StripeMetadataValidPaymentIdNotFoundError,
+} from './stripe.errors'
+
+const logger = createLoggerWithLabel(module)
+
+export const getRedirectUri = () =>
+  `${
+    config.isDev ? 'http://localhost:5001' : config.app.appUrl
+  }/api/v3/payments/stripe/callback`
+
+/**
+ * Extracts the payment id from the metadata field of objects expected to have
+ * it (i.e. payment intents and charges).
+ * @param {Stripe.Metadata} metadata the metadata object which is expected to have a payment id
+ * @returns ok(paymentId) the extracted paymentId
+ * @returns err(StripeMetadataValidPaymentIdNotFoundError) if the payment id was not found or is an invalid BSON object id
+ */
+export const getMetadataPaymentId = (
+  metadata: Stripe.Metadata,
+): Result<string, StripeMetadataValidPaymentIdNotFoundError> => {
+  const logMeta = {
+    action: 'getMetadataPaymentId',
+    metadata,
+  }
+  const paymentId = get(metadata, 'paymentId') // TODO: Extract this value to a constant?
+  if (!paymentId || !mongoose.Types.ObjectId.isValid(paymentId)) {
+    logger.warn({
+      message: 'Got metadata with invalid paymentId',
+      meta: { ...logMeta, metadata },
+    })
+    return err(new StripeMetadataValidPaymentIdNotFoundError())
+  }
+  return ok(paymentId)
+}
+
+/**
+ * State machine that computes the state of the payment, given the list of
+ * Stripe events received via webhooks for this submission.
+ * @param {Stripe.Event[]} events the list of Stripe events for state to be computed on
+ * @returns ok({status, chargeIdLatest}) status and latest charge id based on the list of events
+ * @returns err(ComputePaymentStateError) if there is an error in computing the state
+ */
+export const computePaymentState = (
+  events: Stripe.Event[],
+): Result<
+  Pick<Payment, 'status' | 'chargeIdLatest'>,
+  ComputePaymentStateError
+> => {
+  const logMeta = {
+    action: 'computePaymentState',
+    events,
+  }
+
+  // Helper to get the charge id from a nested charge object.
+  const getChargeId = (charge: string | Stripe.Charge): string =>
+    typeof charge === 'string' ? charge : charge.id
+
+  // We only care about charge and dispute events for computing payment status.
+  // The event types are:
+  // - charge.(failed|pending|refunded|succeeded)
+  // - charge.dispute.(closed|created|funds_reinstated|funds_withdrawn|updated)
+  const chargeEvents = (events as Stripe.DiscriminatedEvent[]).filter(
+    (
+      event: Stripe.DiscriminatedEvent,
+    ): event is
+      | Stripe.DiscriminatedEvent.ChargeEvent
+      | Stripe.DiscriminatedEvent.ChargeDisputeEvent =>
+      event.type === 'charge.failed' ||
+      event.type === 'charge.pending' ||
+      event.type === 'charge.refunded' ||
+      event.type === 'charge.succeeded' ||
+      event.type.startsWith('charge.dispute.'),
```

Comment:
I think this might be terser,
```suggestion
      [
        'charge.failed',
        'charge.pending',
        'charge.refunded',
        'charge.succeeded',
      ].includes(event.type) || event.type.startsWith('charge.dispute.'),
```

Reply:
[3a5c6c8](https://github.com/opengovsg/FormSG/pull/5937/commits/3a5c6c8ee0400042dc3721f7266fd4b977135c80)